From 33acfaf30747592b9fca0913b2c2f8832418d618 Mon Sep 17 00:00:00 2001
From: Jim Stichnoth <jstichnoth@mythtv.org>
Date: Sun, 13 Jul 2014 08:46:16 -0700
Subject: [PATCH 2/2] Make seeks faster for slow decoders / large keyframe
 distances.

Exact seeking for HD-PVR recordings can be notoriously slow because
the keyframe distance is 128 frames and decoding (even GPU hardware
decoding) is often little faster than real-time.

In most cases, it's a much better user experience to make a less
accurate but much faster seek to the preceding keyframe.

There are two changes:

1. Impose a maximum wall-clock time on the forward frame-by-frame
seeking.  After this maximum time elapses, abort any further frame
seeking.

2. In addition, try to predict whether fully accurate seeking will
complete within a reasonable time, and if after a few frames it seems
hopeless, just don't bother seeking any further.

This approach allows fast decoders and video formats amenable to fast
decoding to continue to have exact seeking.

Fixes #12004.
---
 mythtv/libs/libmythtv/avformatdecoder.cpp |   34 ++++++++++++++++++++++++++++-
 mythtv/libs/libmythtv/mythplayer.cpp      |    8 +++++--
 2 files changed, 39 insertions(+), 3 deletions(-)

diff --git a/mythtv/libs/libmythtv/avformatdecoder.cpp b/mythtv/libs/libmythtv/avformatdecoder.cpp
index e9eaad2..64325f8 100644
--- a/mythtv/libs/libmythtv/avformatdecoder.cpp
+++ b/mythtv/libs/libmythtv/avformatdecoder.cpp
@@ -832,7 +832,26 @@ void AvFormatDecoder::SeekReset(long long newKey, uint skipFrames,
     }
 
     // Skip all the desired number of skipFrames
-    for (;skipFrames > 0 && !ateof; skipFrames--)
+
+    // Some seeks can be very slow.  The most common example comes
+    // from HD-PVR recordings, where keyframes are 128 frames apart
+    // and decoding (even hardware decoding) may not be much faster
+    // than realtime, causing some exact seeks to take 2-4 seconds.
+    // If exact seeking is not required, we take some shortcuts.
+    // First, we impose an absolute maximum time we are willing to
+    // spend (maxSeekTimeMs) on the forward frame-by-frame skip.
+    // After that much time has elapsed, we give up and stop the
+    // frame-by-frame seeking.  Second, after skipping a few frames,
+    // we predict whether the situation is hopeless, i.e. the total
+    // skipping would take longer than giveUpPredictionMs, and if so,
+    // stop skipping right away.
+    bool exactSeeks = !GetSeekSnap();
+    const int maxSeekTimeMs = 200;
+    int profileFrames = 0;
+    MythTimer begin(MythTimer::kStartRunning);
+    for (; (skipFrames > 0 && !ateof &&
+            (exactSeeks || begin.elapsed() < maxSeekTimeMs));
+         --skipFrames, ++profileFrames)
     {
         GetFrame(kDecodeVideo);
         if (decoded_video_frame)
@@ -840,6 +859,19 @@ void AvFormatDecoder::SeekReset(long long newKey, uint skipFrames,
             m_parent->DiscardVideoFrame(decoded_video_frame);
             decoded_video_frame = NULL;
         }
+        if (!exactSeeks && profileFrames >= 5 && profileFrames < 10)
+        {
+            const int giveUpPredictionMs = 400;
+            int remainingTimeMs =
+                skipFrames * (float)begin.elapsed() / profileFrames;
+            if (remainingTimeMs > giveUpPredictionMs)
+            {
+              LOG(VB_PLAYBACK, LOG_DEBUG,
+                  QString("Frame-by-frame seeking would take "
+                          "%1 ms to finish, skipping.").arg(remainingTimeMs));
+              break;
+            }
+        }
     }
 
     if (doflush)
diff --git a/mythtv/libs/libmythtv/mythplayer.cpp b/mythtv/libs/libmythtv/mythplayer.cpp
index 4ee83bd..350b15d 100644
--- a/mythtv/libs/libmythtv/mythplayer.cpp
+++ b/mythtv/libs/libmythtv/mythplayer.cpp
@@ -4141,7 +4141,9 @@ bool MythPlayer::HandleProgramEditorActions(QStringList &actions)
             if (seekamount == 0) // 1 frame
                 DoRewind(1, kInaccuracyNone);
             else if (seekamount > 0)
-                DoRewindSecs(seekamount, kInaccuracyEditor, false);
+                // Use fully-accurate seeks for less than 1 second.
+                DoRewindSecs(seekamount, seekamount < 1.0 ? kInaccuracyNone :
+                             kInaccuracyEditor, false);
             else
                 HandleArbSeek(false);
         }
@@ -4150,7 +4152,9 @@ bool MythPlayer::HandleProgramEditorActions(QStringList &actions)
             if (seekamount == 0) // 1 frame
                 DoFastForward(1, kInaccuracyNone);
             else if (seekamount > 0)
-                DoFastForwardSecs(seekamount, kInaccuracyEditor, false);
+                // Use fully-accurate seeks for less than 1 second.
+                DoFastForwardSecs(seekamount, seekamount < 1.0 ? kInaccuracyNone :
+                             kInaccuracyEditor, false);
             else
                 HandleArbSeek(true);
         }
-- 
1.7.10.2

